src/backings/tree.ts

```diff
@@ -0,0 +1,376 @@
+    return this.type.tree_createProof(this.tree, paths);
+  }
+  getPropertyNames(): (string | number)[] {
+    return this.type.tree_getPropertyNames(this.tree);
+  }
+  [Symbol.iterator](): IterableIterator<ValueOf<T>> {
+    return this.values();
+  }
+  abstract getProperty<P extends keyof T>(property: P): ValueOf<T, P>;
+  abstract setProperty<P extends keyof T>(property: P, value: ValueOf<T, P>): boolean;
+  abstract keys(): IterableIterator<KeyOf>;
+  abstract values(): IterableIterator<ValueOf<T>>;
+  abstract entries(): IterableIterator<[KeyOf, ValueOf<T>]>;
+}
+
+export class BasicArrayTreeValue<T extends ArrayLike<unknown>> extends TreeValue<T> {
+  type: BasicArrayType<T>;
+
+  constructor(type: BasicArrayType<T>, tree: Tree) {
+    super(type, tree);
+    this.type = type;
+  }
+
+  getProperty<P extends keyof T>(property: P): ValueOf<T, P> {
+    return this.type.tree_getProperty(this.tree, property) as ValueOf<T, P>;
+  }
+  setProperty<P extends keyof T>(property: P, value: ValueOf<T, P>): boolean {
+    return this.type.tree_setProperty(this.tree, property as number, value);
+  }
+  *keys(): IterableIterator<KeyOf> {
+    const propNames = this.getPropertyNames();
+    // pop off "length"
+    propNames.pop();
+    yield* propNames;
+  }
+  values(): IterableIterator<ValueOf<T>> {
+    return this.type.tree_iterateValues(this.tree) as IterableIterator<ValueOf<T>>;
+  }
+  *entries(): IterableIterator<[KeyOf, ValueOf<T>]> {
+    const keys = this.getPropertyNames();
+    let i = 0;
+    for (const value of this.values()) {
+      yield [keys[i], value];
+      i++;
+    }
+  }
+}
+
+export class CompositeArrayTreeValue<T extends ArrayLike<unknown>> extends TreeValue<T> {
+  type: CompositeArrayType<T>;
+
+  constructor(type: CompositeArrayType<T>, tree: Tree) {
+    super(type, tree);
+    this.type = type;
+  }
+
+  getProperty<P extends keyof T>(property: P): ValueOf<T, P> {
+    return createTreeBacked(this.type.elementType, this.type.tree_getProperty(this.tree, property) as Tree) as ValueOf<
+      T,
+      P
+    >;
+  }
+  setProperty<P extends keyof T>(property: P, value: ValueOf<T, P>): boolean {
+    return this.type.tree_setProperty(
+      this.tree,
+      property as number,
+      isTreeBacked(value)
+        ? value.tree
+        : this.type.elementType.struct_convertToTree((value as unknown) as CompositeValue)
+    );
+  }
+  *keys(): IterableIterator<KeyOf> {
+    const propNames = this.getPropertyNames();
+    // pop off "length"
+    propNames.pop();
+    yield* propNames;
+  }
+  values(): IterableIterator<ValueOf<T>> {
+    return this.type.tree_iterateValues(this.tree) as IterableIterator<ValueOf<T>>;
+  }
+  *entries(): IterableIterator<[KeyOf, ValueOf<T>]> {
+    const keys = this.getPropertyNames();
+    let i = 0;
+    for (const value of this.values()) {
+      yield [keys[i], value];
+      i++;
+    }
+  }
+}
+
+export class BasicListTreeValue<T extends List<unknown>> extends BasicArrayTreeValue<T> {
```

Comment:
it would be way more readable if each class is in separaet file inside tree directory

Reply:
I changed to separate files, now tests fail due to circular import :( plz help

Reply:
I can maybe look into that tommorow